src/main/java/com/exactpro/th2/mstore/AbstractMessageStore.java

```diff
@@ -59,35 +73,105 @@ public final void handle(T delivery) {
+                } catch (CancellationException e) {
+                    if (logger.isWarnEnabled()) {
+                        logger.warn("{} - future related to {} batch is cancelled", getClass().getSimpleName(), formatStoredMessageBatch(batch, false), e);
+                    }
+                } catch (ExecutionException e) {
+                    if (logger.isWarnEnabled()) {
+                        logger.warn("{} - storing {} batch is failure", getClass().getSimpleName(), formatStoredMessageBatch(batch, false), e);
+                    }
+                } catch (TimeoutException e) {
+                    if (logger.isErrorEnabled()) {
+                        logger.error("{} - future related to {} batch can't be complited", getClass().getSimpleName(), formatStoredMessageBatch(batch, false), e);
+                    }
+                    future.cancel(false);
+                }
+            });
+            asyncStoreFutures.keySet().removeAll(futuresToRemove);
+        }
+    }
+
+    protected void storeMessages(List<M> messagesList) throws CradleStorageException {
         logger.debug("Process {} messages started, max {}", messagesList.size(), MAX_MESSAGES_COUNT);
 
         StoredMessageBatch storedMessageBatch = new StoredMessageBatch();
         for (M message : messagesList) {
             MessageToStore messageToStore = convert(message);
             storedMessageBatch.addMessage(messageToStore);
         }
-        store(getCradleManager(), storedMessageBatch);
-        logger.debug("Message Batch stored: stream '{}', direction '{}', id '{}', size '{}', messages '{}'",
-                storedMessageBatch.getStreamName(), storedMessageBatch.getId().getDirection(), storedMessageBatch.getId().getIndex(),
-                storedMessageBatch.getMessageCount(), storedMessageBatch.getMessages());
+        CompletableFuture<Void> future = store(storedMessageBatch);
+        asyncStoreFutures.put(future, storedMessageBatch);
+        future.whenCompleteAsync((value, exception) -> {
+            try {
+                if (exception == null) {
+                    if (logger.isDebugEnabled()) {
+                        logger.debug("{} - batch stored: {}", getClass().getSimpleName(), formatStoredMessageBatch(storedMessageBatch, true));
+                    }
+                } else {
+                    if (logger.isErrorEnabled()) {
+                        logger.error("{} - batch storing is failure: {}", getClass().getSimpleName(), formatStoredMessageBatch(storedMessageBatch, true), exception);
+                    }
+                }
+            } finally {
+                if(asyncStoreFutures.remove(future) == null) {
+                    if (logger.isWarnEnabled()) {
+                        logger.warn("{} - future related to batch {} already removed", getClass().getSimpleName(), formatStoredMessageBatch(storedMessageBatch, true));
+                    }
+                }
+            }
```

Comment:
I think we should extract this into a separate method. It will make the code more readable.